ENH: simultaneous confidence intervals for multinomial proportions (rebased)

wehlutyk opened the pull request:
This should (at least partially) close #2895 (thanks for the help in that discussion!), implementing the "Goodman" and "Sison & Glaz" methods. No existing code was used for the implementation (aside from the MultinomialCI R package for generating test values and finding bugs by comparing intermediate values).

A few notes:

* I just found out about the [informal commit format standard](http://statsmodels.sourceforge.net/devel/dev/maintainer_notes.html#commit-comments), so commit messages don't all follow it; sorry about that (although rebasing+squashing could solve this I believe). I also tried to follow the general code style, but there might be fixes to do (e.g. I define several local functions for `sison_glaz`).
* I didn't include this addition in `docs/source/release/versionX.X.rst`, I'm not sure if it qualifies for it.
* I'm having a few problems building the docs: see #2952 for this
* Another method that might be interesting to implement is [Hou, Chiang, & Tai (2003)](https://www.researchgate.net/profile/Jeng_Tung_Chiang/publication/4897839_A_family_of_simultaneous_confidence_intervals_for_multinomial_proportions/links/55fa704108aeafc8ac39f24c.pdf), using monte-carlo simulations to find more accurate CIs. But it's too much work for me right now.
* Another nice feature I did not implement would be some advice (in the form of warnings) on which method to use depending on what `counts` looks like (e.g., very high number of categories ➔ use `sison_glaz`). But I'd need some help in determining the exact advice to give.
* Performance can be easily improved if necessary (e.g. with a cache on `nu(c)`).

josef-pkt commented:
very good, thanks
I'll go over it in the next few days.

I tried this also in PR #2932 but going the long generic way which didn't work out so well.
But I have MonteCarlo box probability and p-values included, but not yet MonteCarlo confidence intervals.

wehlutyk commented:
Great, thanks! the monte-carlo methods look really neat, great that you already have part of it.

josef-pkt commented:
a few general comments:

based on some comments Sison and Glaz is mainly good if the proportions are roughly equal. The width of the confidence interval is independent of the proportion, and therefore also of the variance.

One thought is to extend or correct asymptotic confidence intervals with approximately exact coverage, so the joint confidence region would also depend on the variance and not be equal width. 
I didn't look at the Hu et all 2003 article yet.

One possibility to add is monte carlo calculation for the probability, so it's possible to check whether the edgeworth based approximation is good for specific cases. I started to do that but for bug hunting because my implementation didn't work correctly, and might still be incorrect in corner cases.
I was running the MC probabilities in batches up to a million replications to see the accuracy at several decimals. 50000 replication is fast an good for a quick cross-check.

josef-pkt commented:
about the monte carlo:

I have currently two functions, one for box probability, and one that uses a call back which I used with scipy.stats.chisquare rejection boolean for p-values. This can be used also for other statistics like Cressie-Read divergence which is also in scipy.stats (and an older version in statsmodels)
These MC functions are good for calculating probabilities once, i.e. we need only a single run through all the replications.

For confidence intervals I think we have to store the monte carlo results so we can search for the bounds. This will be more difficult to do in batches without larger memory requirement. I guess what should work is to keep only the largest and smallest alpha or similar fraction of replications. across batches.

wehlutyk commented:
About this:

> based on some comments Sison and Glaz is mainly good if the proportions are roughly equal. The width of the confidence interval is independent of the proportion, and therefore also of the variance.

I'd like to make sure the advice in the docs is good: is it that the method is more accurate when the proportions are roughly equal (I also read that but couldn't find justification for it in the paper), or is it that, since the widths produced are all equal, it's most *useful* when the proportions are roughly equal?

josef-pkt commented:
It's now 12 days, and I looked at many other things in the meantime, so I'm not really focused on the details anymore.

The function looks good and straight to the point of providing the confidence intervals. Unit test coverage looks also good.
I won't be able to get into this for at least another week and would prefer to merge it as is, than let it wait longer because it implements two standard methods.

@wehlutyk  I also saw your issue/PR for Julia. I guess there are still small sample cases with counts close to zero where neither goodman nor sison-glaz will be very accurate.
I read some articles for small sample confidence regions, but those regions are elliptical or have other difficult shapes that cannot be easily summarized by confidence intervals. My best guess right now is that parametric bootstrap/monte carlo will be the best for those cases. However, I didn't manage yet to finish my PR for that.

I still don't have any better or more precise recommendation for the choice of method than sison-glaz is appropriate when the proportions are similar in magnitude, based on what I read.

Do you think we can merge as is, or is there a part or case that I should specifically look at?

About location and `api`: I think it fits well in `proportions` for users, but if there is more extra (largely internal) code, then I think we can keep it in an internal modul `_multinomial` as I started in my PR.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+
+    Returns
+    -------
+    array of [float, float] arrays
```
should start with some descriptive name similar to Parameters, e.g.
confint : ndarray

josef-pkt replied:
proportion_confint returns a tuple, What's our convention for the shape/type of confint?

wehlutyk replied:
* `confint : ndarray`: ok, will do
* shape/type of confint: since `proportion_confint` returns a tuple, I returned an array of such confints. It can be an array of tuples to be closer to that conventions, or (as now) a 2D ndarray which is more efficient. You tell me.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+
+    Parameters
+    ----------
+    counts : array_like of int
```
int is not required, just `array_like, 1-D` is enough

wehlutyk replied:
Actually, I ran into some errors with the sison-glaz method when not using int (it can't find a value for `c`). I think it makes sense as the method is really only defined for ints.

Should we investigate? Or say `array_like of int, 1-D`?

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+    equal to 5. There is no condition on the number of categories for this
+    method.
+
+    The `sison_glaz` method [3]_ is based on a normal approximation that
```
It's not really a normal approximation, is it?

wehlutyk replied:
Talking about what allows them to go from eq. 6 to eq. 7 (page 2 of the Sison & Glaz paper). It's a "Law of large numbers" approximation ("... [the two distributions] converge to the same multivariate normal distribution ..."), so not sure how to call it. Could be "a distribution approximation that converges ...".

wehlutyk replied:
"based on an approximation of distributions that converges ..."

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+    intervals closer to the desired significance level), but is valid when the
+    number of categories is large. In their paper, Sison & Glaz demo their
+    method with at least 7 categories, so ``len(counts) >= 7`` can be used as a
+    rule of thumb.
```
I would mention that sison-glaz uses uniform width of the confidence interval except at the boundary.

wehlutyk replied:
Noted, adding this.

wehlutyk replied:
I have:

```
    The `sison_glaz` method [3]_ is based on an approximation of distributions
    that converges when the number of observations (``sum(counts)``) goes to
    infinity (likely valid under the same conditions as the `goodman` method,
    i.e. all `counts` values at or above 5), *and* an Edgeworth expansion that
    converges when the number of categories goes to infinity. This method is
    less conservative than the `goodman` method (i.e. it will yield confidence
    intervals closer to the desired significance level), but is valid when the
    number of categories is large. In their paper, Sison & Glaz demo their
    method with at least 7 categories, so ``len(counts) >= 7`` can be used as a
    rule of thumb. Note that this method produces confidence intervals of
    uniform width over all categories (except when the intervals reach 0 or 1,
    in which case they are truncated).
```

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+        raise ValueError('counts must be >= 0')
+    n = counts.sum()
+    k = len(counts)
+    estimated_parameters = counts / n
```
i would use proportions instead of estimated_parameters

wehlutyk replied:
Noted

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+            """Compute mu_r, the r-th factorial moment of a poisson random
+            variable of parameter `l` truncated to `interval = (b, a)`."""
+            b, a = interval
+            return l ** r * (1 - ((poisson_interval((a - r + 1, a), l) -
```
`l` is not a good name because it looks too much like 1
I was reading 1**r

wehlutyk replied:
Right, changed to `p`. (Although I remember this problems appears pretty often, GitHub's rendering on my browser is particularly bad at this (might be my available fonts). In my terminal `1` and `l` look really different.)

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+            (Sison & Glaz's formula (1))."""
+            return (np.prod([poisson_interval(interval, l) for (interval, l)
+                             in zip(intervals, counts)]) *
+                    edgeworth(intervals) / stats.poisson.pmf(n, n))
```
written this way it looks likes product and division
I would expect that using exp sum log would be numerically more stable.

wehlutyk replied:
Right.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,183 @@ def func(qi):
+        ci_lower = np.array(list(map(lambda x: max(x, 0),
+                                     estimated_parameters - c / n)))
+        ci_upper = np.array(list(map(lambda x: min(x, 1),
+                                     estimated_parameters + (c + 2 * g) / n)))
```
it looks to me we can just use numpy instead of list map, something like
`np.minimum(estimated_parameters + (c + 2 * g) / n, 1)`

wehlutyk replied:
I didn't know about `np.minimum` / `np.maximum`, that's great :)

wehlutyk commented:
@josef-pkt sorry for the lag, I've been swamped with work in the last two weeks. I was going to say I wanted to review the doc a little before you could merge, but now that you've reviewed thoroughly (which is awesome thanks!) I'll do the corrections to it all starting today.

wehlutyk commented:
Don't hesitate for more comments if you want other changes (I'll react faster this time). Otherwise I think it's ready to merge.

josef-pkt commented:
@wehlutyk Thanks, I'll try to go through it tonight, when I have time again, so we can merge it soon.

josef-pkt commented:
(too tired today to finish up checking)

overall changes look good

the only thing to check is whether there is a convention for the return shape or type of confint 
I'm still not sure about the explanation in the notes, but it's fine. Maybe I have time to look at sison-glaz paper again.
(The main point of sison-glaz I think is that it's an (asymptotic) approximation to the exact small sample multinomial distribution/probabilities instead of depending on normal or chisquare asymptotic distribution of a test statistic.)

wehlutyk commented:
Okay, I'll leave #3006 to you, you know the codebase with its conventions a lot better.

About the explanation notes (which I'd like to be good, so let's take the time to get it right): there's two approximations in the paper:
* the edgeworth expansion lets them compute `P(b_i <= X_i <= a_i, i = 1 ... k)` with an approximation that converges with `k` going to infinity (i.e. the number of categories)
* then they can look for an integer `c` that make intervals at the right confidence level. But to do that in theory they need access to the real `p_i` (eq. (6)), and so to get around that they approximate eq. (6) with eq. (7), an approximation that stands on a Central Limit Theorem-like result, converging for `n` going to infinity (so number of observations)

This is what I meant in the explanation, admittedly not very clear (here either I guess).

josef-pkt commented:
shape of confint is good, it's the same as `OLS(...).fit().conf_int()`. So no convention problem here (which doesn't rule out inconsistent pattern in other stats, especially with respect to unpacking into lower, upper)

fails if there are zero counts  (It's not a reason to block merge, I don't think I have seen mention of that case in the articles, in some cases for contingency tables there is an option to make counts/proportions stay away from boundaries)

```
import statsmodels.stats.proportion as smprop

methods = ['goodman', 'sison_glaz']

p = [56, 72, 73, 59, 62, 87, 58]
p = [56, 0.5, 73, 59, 62, 87, 58]
p = [1, 1, 3, 2, 4]
#p = [1, 0, 3, 2, 4]  #error "Exception: Couldn't find a value for `c` ..."
for method in methods:
    ci = smprop.multinomial_proportions_confint(p, 0.05, method=method)
    print(ci)
    print()
```

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -123,6 +123,185 @@ def func(qi):
+        Number of observations in each category.
+    alpha : float in (0, 1), optional
+        Significance level, defaults to 0.05.
+    method : {'goodman', 'sison_glaz'}, optional
```
use dash not underline in options `sison-glaz`  
I think that is more standard, and I had it misspelled when I tried an example

wehlutyk replied:
I would've preferred, but chose not to when I saw the methods in [`statsmodels.stats.proportion.proportion_confint`](http://statsmodels.sourceforge.net/stable/generated/statsmodels.stats.proportion.proportion_confint.html) use underscore. Final decision?

josef-pkt replied:
My mistake in proportion_confint, or no convention, changing convention in the past.

compromise: use `sison-glaz' in the docstring
but only check whether method name starts with "sison".
In some cases, I also allowed short cuts or multiple names for the keywords, when there is no unique favorite for the exact spelling. (alternative='two-sided'  '2s' ... IIRC)

wehlutyk commented:
About the failure with proportion 0: I'll look into that. Can't find any reason that requires the counts not to be 0 in the paper, so it should work.

josef-pkt commented:
one name change sison dash glaz

about the convergence:
I never quite understood the detour before (7) when going from (6) to (8). AFAIU, this relates to the usage of parameter estimates p^hat, and not to the distribution itself. In equation (7) the probabilities are evaluated using the approximation in (3) but using the estimated MLE parameters.
I went through the question what parameters to use and whether we are able to remove *nuisance* parameters from the testing in the context of simpler cases for rates and proportions.

So the large sample part, n -> inf, before equation (7) is the justification to use the MLE p^hat.
AFAIU.

I think this is also a reason that it doesn't work when we have zero counts, because then MLE estimate p^hat is zero. 
related aside: in simple proportion confidence intervals the use of MLE parameters corresponds to Wald intervals, and the Wald confidence interval for zero count is just the point 0, i.e. confint = [0, 0]. And that's a reason to special case it or to increase the minimum count to 0.5 or 1.

josef-pkt commented:
to finish up: I would rewrite the notes to mention the approximation to the exact multinomial probability and that the probability is evaluated using the maximum likelihood estimate of the proportions. 
That would avoid getting into the details for justifying MLE convergence.

Also this needs a rebase before merging.

wehlutyk commented:
Right, so:
* added a 0-valued counts check, which raises an exception with proper explanation and suggestion
* fixed the method name handling so only check if it starts with 'sison'
* fixed notes about sison-glaz method, mentioning approximation+MLE, and convergence conditions
* rebased and squashed

I think we're good to go now, unless I missed anything!

wehlutyk commented:
(pending completion of the tests of course)

wehlutyk commented:
Any chance we can merge this? I think all concerns have been addressed :)

josef-pkt commented:
We can merge it into master (for 0.9). 
I added the backport tag as reminder for me to go over it (in the next few days).
Because it's all new functionality, there is no serious reason not to still add it to 0.8

wehlutyk commented:
Great, thanks!

josef-pkt commented:
I checked out the branch that I had before, from 2016-06-18 05:55:31, and I don't see the problem with zero counts/proportions in that case anymore.
Also `goodman` should always work with zero counts.
So, I think raising an exception is too strong.

(I haven't tried the current version of this PR yet.)

for `p = [1, 0, 3, 2, 4]` I get

```
goodman
[[ 0.0118515   0.50723177]
 [ 0.          0.39885409]
 [ 0.07956632  0.67997532]
 [ 0.04013115  0.5991813 ]
 [ 0.12793056  0.75184026]]

sison_glaz
[[ 0.         0.4794962]
 [ 0.         0.3794962]
 [ 0.1        0.6794962]
 [ 0.         0.5794962]
 [ 0.2        0.7794962]]
```

josef-pkt commented:
Because sison-glatz uses uniform width, there shouldn't be inherent problems with 0 counts.
Wald based intervals, or intervals that use the MLE, which is zero for zero counts, will have problems, including my bootstrap version, IIRC, (at least if it is based on wald interval shapes).

(BTW: I might be using now more recent versions of dependencies, numpy, scipy, than I did in June.)

josef-pkt commented:
The rootfinding problem is independent of the zero count

```

---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-8-c179edb7a491> in <module>()
----> 1 smprop.multinomial_proportions_confint([1, 0.5, 3, 2, 4], 0.05, method='sison_glaz')

m:\josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\stats\proportion.py in multinomial_proportions_confint(counts, alpha, method)
    292         while not (nu(c) <= (1 - alpha) < nu(c + 1)):
    293             if c > n:
--> 294                 raise Exception("Couldn't find a value for `c` that "
    295                                 "solves nu(c) <= 1 - alpha < nu(c + 1)")
    296             c += 1

Exception: Couldn't find a value for `c` that solves nu(c) <= 1 - alpha < nu(c + 1)
```

In the old version (from previous comment), the rootfinding worked for the zero count but not for a small positive count.

eg. 
`smprop.multinomial_proportions_confint([56, 0, 73, 59, 62, 87, 58], 0.05, method='sison_glaz')` works and gives result identical to R with verbose=TRUE for `"LOWER(C+1)" "UPPER(C+1)"` but R reports slightly shifted values, which are `"LOWER(SG)"  "UPPER(SG)"` in verbose mode.
smprop.multinomial_proportions_confint([56, 0.5, 73, 59, 62, 87, 58], 0.05, method='sison_glaz') raised an exception in the old version, but succeeds in R.

`smprop.multinomial_proportions_confint([56, 1, 73, 59, 62, 87, 58], 0.05, method='sison_glaz')`  succeeds in old and new version, and is close to R (lower confint at 8 decimals ( which is my print precision) and at around 3 to 4 decimals for upper confint)

josef-pkt commented:
my conclusion for this: 
we should not raise an exception, and just add a comment that there are (still) numerical problems with zero counts or counts < 1.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -120,6 +120,190 @@ def func(qi):
+        # Find the value of `c` that will give us the confidence intervals
+        # (solving nu(c) <= 1 - alpha < nu(c + 1).
+        c = 1.0
+        while not (nu(c) <= (1 - alpha) < nu(c + 1)):
```
you are doing double work here for calculating nu.
when this goes through the loop, then nu(x) is calculated once for `c+1` when c = x-1 and then again for nu(c) when c=x.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -120,6 +120,190 @@ def func(qi):
+            """Compute interval coverage for a given `c` (Sison & Glaz's
+            formula (7))."""
+            return approximated_multinomial_interval(
+                [(count - c, count + c) for count in counts])
```
should this have a protection that `count - c >= 0` and `count + c <= n`
i.e. np.maximum(count - c, 0) and np.minimum(count + c, n)

or is this handled in all the individual parts of the function?

josef-pkt commented:
another observation: sison-glaz doesn't work for fractional counts. If counts are not integer values, then nu(c) is always inf (adding a debug print), so the `while` condition is never satisfied and raises. (int dtype is not required)

josef-pkt commented:
OK, got it:  scipy.stats.poisson.pmf returns zero for decimal values but scipy.stats.poisson._pmf does not
using it in the last line of approximated_multinomial_interval(intervals) makes it work for decimal counts.
`np.log(stats.poisson._pmf(n, n))`  but that doesn't have a check for `n >= 0`

```
smprop.multinomial_proportions_confint([56, 0.001, 73, 59, 62, 87, 58], 0.05, method='sison_glaz')

array([[ 0.09367065,  0.18988168],
       [ 0.        ,  0.04811242],
       [ 0.13670851,  0.23291955],
       [ 0.10126557,  0.1974766 ],
       [ 0.10886048,  0.20507152],
       [ 0.17215146,  0.2683625 ],
       [ 0.09873393,  0.19494496]])
```

which is again the same as R `"LOWER(C+1)" "UPPER(C+1)"` but shifted compared to the reported result in R

josef-pkt commented:
an alternative adjustment is to force counts to be integer valued in the call to
```
approximated_multinomial_interval(
                [(np.maximum(count - c, 0), np.minimum(count + c, n)) 
                 for count in counts])
```
so we evaluate the interval for integer counts, even when the user observed "counts" are not integer valued. This will "round" (down) the interval for the probability calculation.
I'm not sure that's good.

josef-pkt commented:
R multinomialCI seems to have a discontinuity, that we don't have, and causes a discrepancy when one count gets close to zero

ci change slowly except for the transition 0.0563 to 0.0562 (3rd decimal, that's where we disagree)
result in my current version stays at around 0.09367089 for the first value with 1e-6
`ci_0 = smprop.multinomial_proportions_confint([56, 1e-6, 73, 59, 62, 87, 58], 0.05, method='sison_glaz')`

```
> multinomialCI(c(56, 0.0564, 73, 59, 62, 87, 58), 0.05, verbose=FALSE)
           [,1]       [,2]
[1,] 0.09365751 0.18984637
[2,] 0.00000000 0.04823723
[3,] 0.13668934 0.23287820
[4,] 0.10125137 0.19744023
[5,] 0.10884522 0.20503408
[6,] 0.17212732 0.26831618
[7,] 0.09872008 0.19490894
> multinomialCI(c(56, 0.0563, 73, 59, 62, 87, 58), 0.05, verbose=FALSE)
           [,1]       [,2]
[1,] 0.09365754 0.18984638
[2,] 0.00000000 0.04823695
[3,] 0.13668938 0.23287822
[4,] 0.10125139 0.19744024
[5,] 0.10884525 0.20503409
[6,] 0.17212737 0.26831621
[7,] 0.09872011 0.19490895
> multinomialCI(c(56, 0.0562, 73, 59, 62, 87, 58), 0.05, verbose=FALSE)
           [,1]       [,2]
[1,] 0.09618885 0.19237768
[2,] 0.00000000 0.05076796
[3,] 0.13922070 0.23540953
[4,] 0.10378270 0.19997154
[5,] 0.11137656 0.20756539
[6,] 0.17465869 0.27084753
[7,] 0.10125142 0.19744025
> multinomialCI(c(56, 0.0561, 73, 59, 62, 87, 58), 0.05, verbose=FALSE)
           [,1]       [,2]
[1,] 0.09618887 0.19237770
[2,] 0.00000000 0.05076769
[3,] 0.13922073 0.23540957
[4,] 0.10378273 0.19997156
[5,] 0.11137659 0.20756542
[6,] 0.17465874 0.27084757
[7,] 0.10125144 0.19744028
> 
```

It's nothing to worry about much, just puzzling because for larger counts agreement is very high.

josef-pkt commented:
@wehlutyk   Given that I already have everything open and partially finished, I will make the changes and push to a new PR for merging.

josef-pkt commented:
new PR is #3161

josef-pkt commented:
@wehlutyk  I merged #3161 with some changes

Thanks for the contribution
I'm planning to go back to this topic when I have time for `stats` again.

wehlutyk commented:
@josef-pkt great, thanks for taking over! Very happy to see it merged (first contribution to scipy ecosystem :) )
